File: graph/py/bipartiteness.py
# Time:  O(V + E)
# Space: O(b^d)
# Has to test every connected component
def is_bipartite(adjList):
    layer = {}
    for node in range(len(adjList)):
        if node not in layer:
            if not self.bfs_alternate_test(adjList, node, layer):
                return False
    return True

def bfs_alternate_test(self, adjList, source, layer):
    front = [source]
    layer[source] = 0
    while front:
        children = []
        for node in front:
            for child in adjList[node]:
                if child in layer and layer[child] == layer[node]:
                    return False
                elif child not in layer:
                    children.append(child)
                    layer[child] = not layer[node]
        front = children
    return True

File: graph/cxx/bipartiteness.cxx
#include <queue>
#include <vector>

static bool bfs_alternate_test(std::vector<std::vector<int> > adjList, int node, std::vector<int> layer)
{
	std::queue<int> frontier;
	frontier.push(node);
	layer[node] = 0;
	while(not frontier.empty())
	{
		int node = frontier.front();
		frontier.pop();
		for(const int& child: adjList[node])
			if(layer[child] == layer[node])
				return false;
			else if(layer[child] == -1)
			{
				frontier.push(child);
				layer[child] = layer[node] ^ 1;
			}
	}
	return true;
}

bool is_bipartite(std::vector<std::vector<int> > adjList)
{
	std::vector<int> layer(adjList.size(), -1);
	for(int node = 0; node < adjList.size(); node++)
	{
		if(layer[node] == -1)
			if(not bfs_alternate_test(adjList, node, layer))
				return false;
	}
	return true;
}
